Validation/HLTrigger/plugins/HLTGenValSource.cc

```diff
@@ -0,0 +1,480 @@
+  std::string dirName_;
+  std::vector<edm::ParameterSet> histConfigs_;
+  std::vector<edm::ParameterSet> histConfigs2D_;
+  std::vector<edm::ParameterSet> binnings_;
+  std::string hltProcessName_;
+
+  // constructing the info string, which will be written to the output file for display of information in the GUI
+  // the string will have a JSON formating, thus starting here with the opening bracket, which will be close directly before saving to the root file
+  std::string infoString_ = "{";
+
+  // histogram collection per path
+  std::vector<HLTGenValHistCollPath> collectionPath_;
+
+  // HLT config provider/getter
+  HLTConfigProvider hltConfig_;
+
+  // some miscellaneous member variables
+  std::vector<std::string> hltPathsToCheck_;
+  std::vector<std::string> hltPaths;
+  std::vector<std::string> hltPathSpecificCuts;
+  double dR2limit_;
+  bool doOnlyLastFilter_;
+};
+
+HLTGenValSource::HLTGenValSource(const edm::ParameterSet& iConfig)
+    : genParticleToken_(consumes<reco::GenParticleCollection>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("genParticles"))),
+      genMETToken_(consumes<reco::GenMETCollection>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("genMET"))),
+      AK4genJetToken_(consumes<reco::GenJetCollection>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("ak4GenJets"))),
+      AK8genJetToken_(consumes<reco::GenJetCollection>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("ak8GenJets"))),
+      trigEventToken_(consumes<trigger::TriggerEvent>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("TrigEvent"))) {
+  // getting the histogram configurations
+  histConfigs_ = iConfig.getParameterSetVector("histConfigs");
+  histConfigs2D_ = iConfig.getParameterSetVector("histConfigs2D");
+  binnings_ = iConfig.getParameterSetVector("binnings");
+
+  // getting all other configurations
+  dirName_ = iConfig.getParameter<std::string>("DQMDirName");
+  objType_ = iConfig.getParameter<std::string>("objType");
+  dR2limit_ = iConfig.getParameter<double>("dR2limit");
+  doOnlyLastFilter_ = iConfig.getParameter<bool>("doOnlyLastFilter");
+  hltProcessName_ = iConfig.getParameter<std::string>("hltProcessName");
+  hltPathsToCheck_ = iConfig.getParameter<std::vector<std::string>>("hltPathsToCheck");
+}
+
+void HLTGenValSource::dqmBeginRun(const edm::Run& iRun, const edm::EventSetup& iSetup) {
+  // writing general information to info JSON
+  auto t = std::time(nullptr);
+  auto tm = *std::localtime(&t);
+
+  // date and time of running this
+  std::ostringstream timeStringStream;
+  timeStringStream << std::put_time(&tm, "%d-%m-%Y %H-%M-%S");
+  auto timeString = timeStringStream.str();
+  infoString_ += "\"date & time\":\"" + timeString + "\",";
+
+  // CMSSW version
+  std::stringstream CMSSWversionStream(std::getenv("CMSSW_BASE"));
```

Comment:
two things

1) start with lower case so cmsswVersion
2) is there a better way to the CMSSW version. I feel there should be but perhaps not.